contrib/pmsnare/pmsnare.c

```diff
@@ -81,94 +84,93 @@ CODESTARTisCompatibleWithFeature
-
-	*/
+	/* Check if this message is of the type we handle in this parser.
+	 *
+	 * - Find out if the first separator is a tab or a space.
+	 * - If tab-separated, see if the second word is one of our expected tags.
+	 *   - If so, flag as Snare and replace the first tab with space so that hostname and syslog tag parse properly
+	 *   - Else not a snare message.
+	 * - Else assume valid 3164 timestamp, move over to the syslog tag.
+	 * - See if that is tab-separated from one of our expected tags.
+	 *   - If so, flag as Snare.
+	 * - See if either type flagged as Snare.
+	 *   - If so, replace the tab with a space so that it will be parsed properly.
+	 */
+	 
 	snaremessage=0;
-	lenMsg = pMsg->iLenRawMsg - pMsg->offAfterPRI;
 	/* note: offAfterPRI is already the number of PRI chars (do not add one!) */
-	p2parse = pMsg->pszRawMsg + pMsg->offAfterPRI; /* point to start of text, after PRI */
+	lenMsg = pMsg->iLenRawMsg - pMsg->offAfterPRI;
+	/* point to start of text, after PRI */
+	p2parse = pMsg->pszRawMsg + pMsg->offAfterPRI;
 	dbgprintf("pmsnare: msg to look at: [%d]'%s'\n", lenMsg, p2parse);
 	if((unsigned) lenMsg < 30) {
 		/* too short, can not be "our" message */
-		dbgprintf("msg too short!\n");
+		dbgprintf("Message is too short to be Snare!\n");
 		ABORT_FINALIZE(RS_RET_COULD_NOT_PARSE);
 	}
 
-	while(lenMsg && *p2parse != ' ' && *p2parse != '\t' && *p2parse != '#') {
+	/* Check for non-syslog snare format. */
+	while(lenMsg && *p2parse != ' ' && *p2parse != tabRepresentation[0]) {
+		/* Ignore anything before first space/tab/special. */
 		--lenMsg;
 		++p2parse;
 	}
-	dbgprintf("pmsnare: separator [%d]'%s'  msg after the first separator: [%d]'%s'\n", tablength,
-	TabRepresentation,lenMsg, p2parse);
-	if ((lenMsg > tablength) && (*p2parse == '\t' || strncasecmp((char*) p2parse, TabRepresentation ,
-	tablength-1) == 0)) {
-	//if ((lenMsg > tablength) && (*p2parse == '\t' || *p2parse == '#')) {
-	dbgprintf("pmsnare: tab separated message\n");
-		if(strncasecmp((char*) (p2parse + tablength - 1), "MSWinEventLog", 13) == 0) {
-			snaremessage=13; /* 0 means not a snare message, a number is how long the tag is */
+	if ((lenMsg > tabLength) && (strncasecmp((char *)p2parse, tabRepresentation, tabLength) == 0)) {
+		/* The first separator is a tab. */
+		dbgprintf("pmsnare: tab separated message\n");
+		dbgprintf("pmsnare: tab [%d]'%s'	msg at the first separator: [%d]'%s'\n",
+			tabLength, tabRepresentation, lenMsg, p2parse);
+
+		/* Look for the Snare tag. */
+		if(strncasecmp((char*)(p2parse + tabLength), "MSWinEventLog", 13) == 0) {
+			dbgprintf("Found a non-syslog Windows Snare message.\n");
+			snaremessage = p2parse - pMsg->pszRawMsg + tabLength + 13;
 		}
-		if(strncasecmp((char*) (p2parse + tablength - 1), "LinuxKAudit", 11) == 0) {
-			snaremessage=11; /* 0 means not a snare message, a number is how long the tag is */
+		else if(strncasecmp((char*) (p2parse + tabLength), "LinuxKAudit", 11) == 0) {
```

Comment:
tablength is a string with a terminating \0 on it, so "#011" is a length 5. "\t" is length 2

This is why the old code used tablength -1 for comparison and tablength -2 for shrinking the message